File: output/cursor/java/medium/q17-longest-substring-without-repeating-characters.java
/** https://leetcode.com/problems/longest-substring-without-repeating-characters */
// Problem: Longest Substring Without Repeating Characters

//Problem Description:
//Given a string s, find the length of the longest substring without duplicate characters.

// Constraints:
// 0 <= s.length <= 5 * 10^4
// s consists of English letters, digits, symbols and spaces.

//Code Structure:

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0, right = 0;
        int maxLen = 0;
        Set<Character> seen = new HashSet<>();
        
        while (right < s.length()) {
            if (!seen.contains(s.charAt(right))) {
                seen.add(s.charAt(right));
                right++;
                maxLen = Math.max(maxLen, right - left);
            } else {
                seen.remove(s.charAt(left));
                left++;
            }
        }
        
        return maxLen;
    }
}
    

File: output/amazon_q/python/medium/q17-longest-substring-without-repeating-characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/
# Problem: Longest Substring Without Repeating Characters

#Problem Description:
#Given a string s, find the length of the longest substring without duplicate characters.

# Constraints:
# 0 <= s.length <= 5 * 10^4
# s consists of English letters, digits, symbols and spaces.

#Code Structure:

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        char_set = set()
        left = 0
        max_length = 0

        for right in range(len(s)):
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1
            char_set.add(s[right])
            max_length = max(max_length, right - left + 1)

        return max_length                                                                    
    